### PR TITLE
feat: Transformed pandas processing into SQL for leaderboard

### PR DESCRIPTION
**Pull request by irfanuddinahmad**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-analytics-data-api doesn't install it
    - `test-master.in` if edx-analytics-data-api pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the edx-analytics-data-api requirements installed or if you checked out edx-enterprise-data into the src directory used by edx-analytics-data-api, you can run this command through an edx-analytics-data-api shell.
        - It would be `./manage.py makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise-data/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise-data/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise-data/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise-data), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise-data/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-analytics-data-api](https://github.com/openedx/edx-analytics-data-api) to upgrade dependencies (including edx-enterprise-data)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-analytics-data-api will look for the latest version in PyPi.
    - Note: the edx-enterprise-data constraint in edx-analytics-data-api **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise_data/admin_analytics/database/queries/fact_engagement_admin_dash.py [hunk 1]
```diff
@@ -19,3 +19,43 @@ def get_learning_hours_and_daily_sessions_query():
...
+            SELECT
+                Engagement.email,
+                learning_time_hours,
+                sessions,
+                average_session_length,
+                course_completions
+            FROM Engagement
+            LEFT JOIN Completions
+            ON Engagement.email = Completions.email
+            ORDER BY learning_time_hours desc, sessions desc, course_completions desc
```
Comment (muhammad-ammar) [inline enterprise_data/admin_analytics/database/queries/fact_engagement_admin_dash.py]: Facing the below error locally on read replica when I execute the query in mysql shell.

```
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '== 0 THEN 0.0
            ELSE ROUND(learning_time_hours / sessions::float, 1)
```

File: enterprise_data/api/v1/views/analytics_leaderboard.py [hunk 2]
```diff
@@ -36,61 +30,17 @@ class AdvanceAnalyticsLeaderboardView(APIView):
...
-
-        leaderboard_df = engage.merge(completions, on="email", how="left")
-        leaderboard_df = leaderboard_df.sort_values(
-            by=["learning_time_hours", "daily_sessions", "course_completions"],
-            ascending=[False, False, False],
-        )
-
+        response_type = request.query_params.get('response_type', ResponseType.JSON.value)
+        leaderboard = FactEngagementAdminDashTable().get_leaderboard(enterprise_uuid, start_date, end_date)
+        leaderboard_df = pd.DataFrame(leaderboard)
```
Comment (muhammad-ammar) [inline enterprise_data/api/v1/views/analytics_leaderboard.py]: Why are we again converting the results to dataframe?

Comment (muhammad-ammar) [reply enterprise_data/api/v1/views/analytics_leaderboard.py]: I believe we can convert the below logic into python without using pandas dataframe. We have planned to remove the pandas once we convert all the code to mysql queries.

```
# move the aggregated row with email 'null' to the end of the table
idx = leaderboard_df.index[leaderboard_df['email'] == 'null']
leaderboard_df.loc[idx, 'email'] = 'learners who have not shared consent'
leaderboard_df = pd.concat([leaderboard_df.drop(idx), leaderboard_df.loc[idx]])
```

Comment (irfanuddinahmad) [reply enterprise_data/api/v1/views/analytics_leaderboard.py]: pandas is python also!

Comment (muhammad-ammar) [reply enterprise_data/api/v1/views/analytics_leaderboard.py]: >  We can convert the below logic into python **without using pandas dataframe**.

Comment (irfanuddinahmad) [reply enterprise_data/api/v1/views/analytics_leaderboard.py]: I think pandas might be more efficient in this processing ...

**Review by muhammad-ammar — APPROVED** [review]

**Pull request merged**